packages/react-reconciler/src/ReactFiberWorkLoop.old.js

```diff
@@ -838,45 +839,58 @@ function performConcurrentWorkOnRoot(root, didTimeout) {
       throw fatalError;
     }
 
-    // Check if this render may have yielded to a concurrent event, and if so,
-    // confirm that any newly rendered stores are consistent.
-    // TODO: It's possible that even a concurrent render may never have yielded
-    // to the main thread, if it was fast enough, or if it expired. We could
-    // skip the consistency check in that case, too.
-    const renderWasConcurrent = !includesBlockingLane(root, lanes);
-    const finishedWork: Fiber = (root.current.alternate: any);
-    if (
-      renderWasConcurrent &&
-      !isRenderConsistentWithExternalStores(finishedWork)
-    ) {
-      // A store was mutated in an interleaved event. Render again,
-      // synchronously, to block further mutations.
-      exitStatus = renderRootSync(root, lanes);
-
-      // We need to check again if something threw
-      if (exitStatus === RootErrored) {
-        const errorRetryLanes = getLanesToRetrySynchronouslyOnError(root);
-        if (errorRetryLanes !== NoLanes) {
-          lanes = errorRetryLanes;
-          exitStatus = recoverFromConcurrentError(root, errorRetryLanes);
-          // We assume the tree is now consistent because we didn't yield to any
-          // concurrent events.
+    if (exitStatus === RootDidNotComplete) {
+      // The render unwound without completing the tree. This happens in special
+      // cases where need to exit the current render without producing a
+      // consistent tree or committing.
+      //
+      // This should only happen during a concurrent render, not a discrete or
+      // synchronous update. We should have already checked for this when we
+      // unwound the stack.
```

Comment:
Just noting here for reference when we follow on the task. This doesn't happen atm. At least in a synchronous forced hydration if something suspends, we can end up with this status and subsequently throw.